crates/iceberg/src/writer/file_writer/location_generator.rs

```diff
@@ -0,0 +1,99 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains the location generator and file name generator for generating path of data file.
+
+use std::sync::{atomic::AtomicU64, Arc};
+
+use crate::spec::DataFileFormat;
+
+/// `LocationGenerator` used to generate the location of data file.
+pub trait LocationGenerator: Clone + Send + 'static {
```

Comment:
Can we merge `LocationGenerator` and `FileNameGenerator`? Both of them only have one easy function to implement. By merging them, we can reduce an extra generic param for users to use.

Or maybe we can go deeper by generating the final location in one shot.

Reply:
The `LocationGenerator` is to generate the path for data dir, such as `/table/data` or `/table/id=1/data/`. `FileNameGenerator` generates the file name, such as `1-1.parquet` or `1-1-delete.parquet`. The benefit of separating them is that we can let users reuse one part of it. E.g. default `LocationGenerator` and custom `FileNameGenerator`. And that's iceberg-java support. In iceberg-java, [`LocationProvider` ](https://github.com/apache/iceberg/blob/062704283918ae843d0e40b3f279f1482f0835e5/api/src/main/java/org/apache/iceberg/io/LocationProvider.java#L30)is to provide the data dir. [The ` FileNameGenerator` is in OutFileFactory.](https://github.com/apache/iceberg/blob/062704283918ae843d0e40b3f279f1482f0835e5/core/src/main/java/org/apache/iceberg/io/OutputFileFactory.java#L93)